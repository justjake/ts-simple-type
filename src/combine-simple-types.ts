import { isSimpleTypeLiteral, isSimpleTypePrimitive, SimpleType, SimpleTypeInterface, SimpleTypeKind, SimpleTypeLiteral, SimpleTypeMemberNamed, SimpleTypeTuple } from "./simple-type";
import { zip } from "./util";

export function combineSimpleTypes(types: SimpleType[]): SimpleType {
	const combined = new Map<SimpleTypeKind, SimpleType>();

	for (const type of types) {
		if (isSimpleTypeLiteral(type)) {
			setExistingOrCombine(combined, type.kind, type, (existing: SimpleTypeLiteral) => (existing.value === type.value ? existing : { kind: SimpleTypeKind.NEVER }));
		}

		if (isSimpleTypePrimitive(type)) {
			setExistingOrCombine(combined, type.kind, type, existing => existing);
		}

		switch (type.kind) {
			case SimpleTypeKind.CLASS:
				setExistingOrCombine(
					combined,
					SimpleTypeKind.INTERFACE,
					type,
					(existing: SimpleTypeInterface) =>
						({
							kind: SimpleTypeKind.INTERFACE,
							members: combineNamedMembersIntoInterface([...type.properties, ...type.methods, ...(existing.members || [])])
						} as SimpleTypeInterface)
				);
				break;

			case SimpleTypeKind.OBJECT:
			case SimpleTypeKind.INTERFACE:
				setExistingOrCombine(
					combined,
					SimpleTypeKind.INTERFACE,
					type,
					(existing: SimpleTypeInterface) =>
						({
							kind: SimpleTypeKind.INTERFACE,
							members: combineNamedMembersIntoInterface([...(type.members || []), ...(existing.members || [])])
						} as SimpleTypeInterface)
				);
				break;

			case SimpleTypeKind.TUPLE:
				setExistingOrCombine(combined, SimpleTypeKind.TUPLE, type, (existing: SimpleTypeTuple) => {
					const members = zipCombine(type.members, existing.members, ([mA, mB]) => ({
						optional: mA.optional && mB.optional,
						type: combineSimpleTypes([mA.type, mB.type])
					}));

					return {
						...type,
						members: members == null ? [{ type: { kind: SimpleTypeKind.NEVER } }] : members
					} as SimpleTypeTuple;
				});
				break;
		}
	}

	//console.dir(combined, { depth: 4 });
	const combinedTypes = Array.from(combined.values());
	if (combinedTypes.length === 1) {
		return combinedTypes[0];
	}
	/*} else {
		return {
			kind: SimpleTypeKind.INTERSECTION,
			types: combinedTypes
		};
	}*/

	return { kind: SimpleTypeKind.NEVER };
}

function zipCombine<T, U>(listA: T[], listB: U[], combine: (lists: [T, U]) => T | U): (T | U)[] | null {
	const result = zip(listA, listB);
	return result == null ? null : result.map(combine);
}

function setExistingOrCombine<T, U, Z = U>(map: Map<T, U>, key: T, alternative: U, combine: (existing: Z) => U): Map<T, U> {
	const existing = map.get(key) as Z | undefined;
	map.set(key, existing == null ? alternative : combine(existing));
	return map;
}

function combineNamedMembersIntoInterface(members: SimpleTypeMemberNamed[]): SimpleTypeMemberNamed[] {
	return Array.from(
		members
			.reduce(
				(map, member) =>
					setExistingOrCombine(map, member.name, member, existingMember => ({
						name: member.name,
						optional: member.optional && existingMember.optional,
						type: combineSimpleTypes([member.type, existingMember.type])
					})),
				new Map<string, SimpleTypeMemberNamed>()
			)
			.values()
	);
}