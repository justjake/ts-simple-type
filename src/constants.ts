import { Type } from "typescript";
import { SimpleType, SimpleTypeNever, SimpleTypeUnknown } from "./simple-type";

export const DEFAULT_TYPE_CACHE = new WeakMap<Type, SimpleType>();

export const DEFAULT_RESULT_CACHE = new Map<string, WeakMap<SimpleType, WeakMap<SimpleType, boolean>>>();

export const DEFAULT_GENERIC_PARAMETER_TYPE: SimpleTypeUnknown = { kind: "UNKNOWN" };

export const NEVER_TYPE: SimpleTypeNever = { kind: "NEVER" };
