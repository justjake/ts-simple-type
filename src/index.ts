export * from "./simple-type";
export * from "./ts-module";

export * from "./is-assignable/simple-type-comparison-options";
export * from "./is-assignable/is-assignable-to-primitive-type";
export * from "./is-assignable/is-assignable-to-type";
export * from "./is-assignable/is-assignable-to-value";
export * from "./is-assignable/is-assignable-to-simple-type-kind";

export * from "./transform/to-simple-type";
export * from "./transform/type-to-string";
export * from "./transform/serialize-simple-type";

export * from "./utils/validate-type";
