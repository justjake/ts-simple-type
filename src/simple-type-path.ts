import { SimpleType, SimpleTypeFunctionParameter, SimpleTypeMember, SimpleTypeMemberNamed } from "./simple-type";
import { simpleTypeToString } from "./transform/simple-type-to-string";

type SimpleTypePathStepKind =
	| "NAMED_MEMBER"
	| "INDEXED_MEMBER"
	| "STRING_INDEX"
	| "NUMBER_INDEX"
	| "VARIANT" // Union, Intersection, Enum
	| "AWAITED"
	| "TYPE_PARAMETER"
	| "TYPE_PARAMETER_CONSTRAINT"
	| "TYPE_PARAMETER_DEFAULT"
	| "PARAMETER"
	| "RETURN"
	| "GENERIC_ARGUMENT"
	| "CALL_SIGNATURE"
	| "CTOR_SIGNATURE"
	| "GENERIC_TARGET"
	| "ALIASED";

export interface SimpleTypePathStepBase {
	step: SimpleTypePathStepKind;
	from: SimpleType; // TODO: should we really include the "from" type in the path?
}

export interface SimpleTypePathStepNamedMember extends SimpleTypePathStepBase {
	step: "NAMED_MEMBER";
	member: SimpleTypeMemberNamed; // TODO: omit type?
	index: number;
}

export interface SimpleTypePathStepIndexedMember extends SimpleTypePathStepBase {
	step: "INDEXED_MEMBER";
	member: SimpleTypeMember;
	index: number;
}

export interface SimpleTypePathStepStringIndex extends SimpleTypePathStepBase {
	step: "STRING_INDEX";
}

export interface SimpleTypePathStepNumberIndex extends SimpleTypePathStepBase {
	step: "NUMBER_INDEX";
}

/**
 * Variant in a union, intersection, or enum type.
 */
export interface SimpleTypePathStepVariant extends SimpleTypePathStepBase {
	step: "VARIANT";
	index: number;
}

export interface SimpleTypePathStepAwaited extends SimpleTypePathStepBase {
	step: "AWAITED";
}

export interface SimpleTypePathStepTypeParameter extends SimpleTypePathStepBase {
	step: "TYPE_PARAMETER";
	index: number;
	name: string;
}

export interface SimpleTypePathStepTypeParameterConstraint extends SimpleTypePathStepBase {
	step: "TYPE_PARAMETER_CONSTRAINT";
}

export interface SimpleTypePathStepTypeParameterDefault extends SimpleTypePathStepBase {
	step: "TYPE_PARAMETER_DEFAULT";
}

export interface SimpleTypePathStepParameter extends SimpleTypePathStepBase {
	step: "PARAMETER";
	index: number;
	parameter: SimpleTypeFunctionParameter; // TODO: omit type?
}

export interface SimpleTypePathStepReturn extends SimpleTypePathStepBase {
	step: "RETURN";
}

/** Visit a call signature of a interface or class type. Arguably could use "variant", but... */
export interface SimpleTypePathStepCallSignature extends SimpleTypePathStepBase {
	step: "CALL_SIGNATURE";
}

/** Visit a call signature of a interface or class type. Arguably could use "variant", but... */
export interface SimpleTypePathStepCtorSignature extends SimpleTypePathStepBase {
	step: "CTOR_SIGNATURE";
}

/**
 * Step from a generic instantiation (GENERIC_ARGUMENTS) to one of the arguments
 * used for the instantiation.
 *
 * ```typescript
 * interface Foo<T> {
 *   bar: T
 * }
 *
 * //                     vvvvvv GENERIC_ARGUMENT
 * type FooInstance = Foo<string>
 * //                 ^^^^^^^^^^^ GENERIC_ARGUMENTS
 * ```
 */
export interface SimpleTypePathStepGenericArgument extends SimpleTypePathStepBase {
	step: "GENERIC_ARGUMENT";
	index: number;
	name?: string;
}

/**
 * Step from a generic instantiation (GENERIC_ARGUMENTS) to the target of the
 * instantiation.
 *
 * ```typescript
 * interface Foo<T> {
 *   bar: T
 * }
 *
 * //                 vvv GENERIC_TARGET
 * type FooInstance = Foo<string>
 * //                 ^^^^^^^^^^^ GENERIC_ARGUMENTS
 * ```
 */
export interface SimpleTypePathStepGenericTarget extends SimpleTypePathStepBase {
	step: "GENERIC_TARGET";
}

export interface SimpleTypePathStepAliased extends SimpleTypePathStepBase {
	step: "ALIASED";
}

export type SimpleTypePathStep =
	| SimpleTypePathStepNamedMember
	| SimpleTypePathStepIndexedMember
	| SimpleTypePathStepStringIndex
	| SimpleTypePathStepNumberIndex
	| SimpleTypePathStepVariant
	| SimpleTypePathStepAwaited
	| SimpleTypePathStepTypeParameter
	| SimpleTypePathStepTypeParameterConstraint
	| SimpleTypePathStepTypeParameterDefault
	| SimpleTypePathStepCallSignature
	| SimpleTypePathStepCtorSignature
	| SimpleTypePathStepParameter
	| SimpleTypePathStepReturn
	| SimpleTypePathStepGenericArgument
	| SimpleTypePathStepGenericTarget
	| SimpleTypePathStepAliased;

/**
 * Describes a traversal path from a starting type to a destination type.
 *
 * Given this type:
 * ```typescript
 * type Deep = {
 *   foo: {
 *     [key: string]: () => string | number
 *                                // ^^^^^^
 *   }
 * }
 * ```
 *
 * A path from `Deep` to the `number` type highlighted would be:
 *
 * 1. NAMED_MEMBER (name foo)
 * 2. STRING_INDEX
 * 3. RETURN
 * 4. VARIANT (index 1)
 */
export type SimpleTypePath = SimpleTypePathStep[]; // TODO: consider single linked list to save memory?

export const SimpleTypePath = {
	empty(): SimpleTypePath {
		return [] as SimpleTypePath;
	},

	/** @returns true if `path` includes a step from `type`. */
	includes(path: SimpleTypePath, type: SimpleType): boolean {
		return path.some(step => step.from === type);
	},

	/** @returns A subpath from `fromType`, or undefined if not present in the path. */
	getSubpathFrom(path: SimpleTypePath, fromType: SimpleType): SimpleTypePath | undefined {
		const fromIndex = path.findIndex(step => step.from === fromType);
		if (fromIndex < 0) {
			return undefined;
		}

		return path.slice(fromIndex);
	},

	concat(prefix: SimpleTypePath, suffix: SimpleTypePath | SimpleTypePath[number] | undefined): SimpleTypePath {
		return suffix ? prefix.concat(suffix) : prefix.concat();
	},

	last(path: SimpleTypePath): SimpleTypePathStep | undefined {
		return path[path.length - 1];
	},

	toTypescript(path: SimpleTypePath): string {
		const parts: string[] = [];
		if (path.length === 0) {
			parts.push("T");
		}

		for (const step of path) {
			if (parts.length === 0) {
				parts.push(step.from.name ?? "T");
			}

			switch (step.step) {
				case "ALIASED":
					continue;
				case "AWAITED":
					parts.unshift("Awaited<");
					parts.push(">");
					continue;
				case "CALL_SIGNATURE":
				case "CTOR_SIGNATURE":
				case "GENERIC_TARGET":
					continue;
				case "INDEXED_MEMBER":
					parts.push(`[${step.index}]`);
					continue;
				case "NAMED_MEMBER":
					parts.push(`["${step.member.name}"]`);
					continue;
				case "NUMBER_INDEX":
					parts.push(`[number]`);
					continue;
				case "STRING_INDEX":
					parts.push(`[string]`);
					continue;
				case "PARAMETER":
					parts.unshift("Parameters<");
					parts.push(`>[${step.index}]`);
					continue;
				case "TYPE_PARAMETER":
				case "TYPE_PARAMETER_CONSTRAINT":
				case "TYPE_PARAMETER_DEFAULT":
					continue;
				case "VARIANT":
					continue;
				case "RETURN":
					parts.unshift("ReturnType<");
					parts.push(">");
					continue;
				case "GENERIC_ARGUMENT":
					continue;
				default:
					unreachable(step);
			}
		}

		return parts.join("");
	},

	toString(path: SimpleTypePath, target?: SimpleType): string {
		const arrow = (name: string) => `~${name}~>`;
		const typeName = (type: SimpleType) => {
			try {
				return type.name ?? simpleTypeToString(type);
			} catch (error) {
				return `(Error converting type to string: ${error})`;
			}
		};

		const parts: string[] = [];
		if (path.length === 0) {
			parts.push("T");
		}

		for (const step of path) {
			if (parts.length === 0) {
				parts.push(step.from.name ?? "T");
			}

			switch (step.step) {
				case "ALIASED":
					continue;
				case "AWAITED":
					parts.push(arrow("awaited"));
					break;
				case "CALL_SIGNATURE":
					parts.push(arrow("asFunction"));
					break;
				case "CTOR_SIGNATURE":
					parts.push(arrow("asConstructor"));
					break;
				case "GENERIC_ARGUMENT":
					parts.push(`<<${step.name ?? step.index}>>`);
					break;
				case "GENERIC_TARGET":
					parts.push(arrow("instantiatedFrom"));
					break;
				case "INDEXED_MEMBER": {
					const optional = step.member.optional ? "?" : "";
					parts.push(`[${step.index}]${optional}`);
					break;
				}
				case "NAMED_MEMBER": {
					const optional = step.member.optional ? "?" : "";
					// parts.push(`["${step.member.name}"]${optional}`);
					parts.push(`.${step.member.name}${optional}`);
					break;
				}
				case "NUMBER_INDEX": {
					parts.push(`[number]`);
					break;
				}
				case "PARAMETER":
					parts.push(arrow(`(${step.parameter.name ?? step.index})`));
					break;
				case "RETURN":
					parts.push(".()");
					break;
				case "STRING_INDEX":
					parts.push("[string]");
					break;
				case "TYPE_PARAMETER":
					parts.push(`<${step.name ?? step.index}>`);
					break;
				case "TYPE_PARAMETER_CONSTRAINT":
					parts.push(arrow("constraintType"));
					break;
				case "TYPE_PARAMETER_DEFAULT":
					parts.push(arrow("defaultType"));
					break;
				case "VARIANT": {
					switch (step.from.kind) {
						case "UNION":
							parts.push(arrow(`|${step.index}`));
							break;
						case "INTERSECTION":
							parts.push(arrow(`&${step.index}`));
							break;
						case "ENUM":
							parts.push(arrow(`.${step.index}`));
							break;
					}
					break;
				}
				default:
					unreachable(step);
			}
		}

		if (target) {
			parts.push(`: ${typeName(target)}`);
		}

		return parts.join("");
	}
} as const;

interface Foo<A extends Map<string, string>> {
	bar: A;
}

// path to Map<string, string>
// Foo<A extends>

export function unreachable(x: never): never {
	throw new Error(`Should be unreachable, instead exists: ${JSON.stringify(x)}`);
}
