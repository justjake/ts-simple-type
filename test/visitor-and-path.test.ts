import { test, expect } from "@jest/globals"

import { SimpleTypePath } from "../src/simple-type-path"
import { toSimpleType } from "../src/transform/to-simple-type"
import { VisitorArgs, Visitor, walkDepthFirst } from "../src/visitor"

import { getTestTypes } from "./helpers/get-test-types"

const EXAMPLE_TYPES = `
interface GenericInterface<ParamWithConstraint extends Obj1, ParamWithDefault = Obj2> {
  constrained: ParamWithConstraint
  defaulted: ParamWithDefault
}

type Obj1 = {
  table: "block" | "collection" | "activity"
  method: (param: Obj2) => Obj2
}

type Obj2 = {
  object2: true
}

 export interface Haystack {
    hello: {
      world: string
    } & {
      today: [1, 2, 3]
    }
    frog: GenericInterface<{ table: "block" }, { value: string }>
 } 
`

test("visitDepthFirst", () => {
	const { types, typeChecker } = getTestTypes(["Haystack"], EXAMPLE_TYPES)
	const simpleType = toSimpleType(types.Haystack, typeChecker, {
		addMethods: true,
		cache: new WeakMap(),
	})

	const toStrings = (args: VisitorArgs<void>) =>
		`
toString:     ${SimpleTypePath.toString(args.path, args.type)}
toTypescript: ${SimpleTypePath.toTypescript(args.path)}
`.trim()

	const visitBeforeOrder: string[] = []
	const visitAfterOrder: string[] = []

	walkDepthFirst([], simpleType, {
		before(args) {
			visitBeforeOrder.push(toStrings(args))
		},
		after(args) {
			visitBeforeOrder.push(toStrings(args))
		},
	})

	expect(`\n${visitBeforeOrder.join("\n\n")}`).toMatchSnapshot()
	expect(`\n${visitAfterOrder.join("\n\n")}`).toMatchSnapshot()
})

test("visitDepthFirst: makes errors nice", () => {
	const { types, typeChecker } = getTestTypes(["Haystack"], EXAMPLE_TYPES)
	const simpleType = toSimpleType(types.Haystack, typeChecker, {
		addMethods: true,
		cache: new WeakMap(),
	})

	expect(() => {
		walkDepthFirst([], simpleType, {
			before: ({ path }) => {
				if (path.length > 5) {
					throw new Error("oops")
				}
			},
			after: undefined,
		})
	}).toThrowError(/Path:.*$/)
})

test("mapJsonStep", () => {
	const { types, typeChecker } = getTestTypes(["Haystack"], EXAMPLE_TYPES)
	const simpleType = toSimpleType(types.Haystack, typeChecker, {
		addMethods: true,
		cache: new WeakMap(),
	})

	const visitBeforeOrder: string[] = []

	walkDepthFirst([], simpleType, {
		before(args) {
			visitBeforeOrder.push(SimpleTypePath.toString(args.path, args.type))
		},
		after: undefined,
		traverse: Visitor.mapJsonStep,
	})

	expect(visitBeforeOrder).toMatchSnapshot()
})
