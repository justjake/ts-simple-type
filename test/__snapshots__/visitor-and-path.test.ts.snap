// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapJsonStep 1`] = `
[
  "T: Haystack",
  "Haystack.hello: { world: string; } & { today: [1, 2, 3]; }",
  "Haystack.hello~&0~>: { world: string; }",
  "Haystack.hello~&0~>.world: string",
  "Haystack.hello~&1~>: { today: [1, 2, 3]; }",
  "Haystack.hello~&1~>.today: [1, 2, 3]",
  "Haystack.hello~&1~>.today[0]: 1",
  "Haystack.hello~&1~>.today[1]: 2",
  "Haystack.hello~&1~>.today[2]: 3",
  "Haystack.frog: GenericInterface<{ table: "block"; }, { value: string; }>",
  "Haystack.frog: GenericInterface",
  "Haystack.frog.constrained: { table: "block"; }",
  "Haystack.frog.constrained.table: "block"",
  "Haystack.frog.defaulted: { value: string; }",
  "Haystack.frog.defaulted.value: string",
]
`;

exports[`visitDepthFirst 1`] = `
"
toString:     T: Haystack
toTypescript: T

toString:     Haystack.hello: { world: string; } & { today: [1, 2, 3]; }
toTypescript: Haystack["hello"]

toString:     Haystack.hello~&0~>: { world: string; }
toTypescript: Haystack["hello"]

toString:     Haystack.hello~&0~>.world: string
toTypescript: Haystack["hello"]["world"]

toString:     Haystack.hello~&0~>.world: string
toTypescript: Haystack["hello"]["world"]

toString:     Haystack.hello~&0~>: { world: string; }
toTypescript: Haystack["hello"]

toString:     Haystack.hello~&1~>: { today: [1, 2, 3]; }
toTypescript: Haystack["hello"]

toString:     Haystack.hello~&1~>.today: [1, 2, 3]
toTypescript: Haystack["hello"]["today"]

toString:     Haystack.hello~&1~>.today[0]: 1
toTypescript: Haystack["hello"]["today"][0]

toString:     Haystack.hello~&1~>.today[0]: 1
toTypescript: Haystack["hello"]["today"][0]

toString:     Haystack.hello~&1~>.today[1]: 2
toTypescript: Haystack["hello"]["today"][1]

toString:     Haystack.hello~&1~>.today[1]: 2
toTypescript: Haystack["hello"]["today"][1]

toString:     Haystack.hello~&1~>.today[2]: 3
toTypescript: Haystack["hello"]["today"][2]

toString:     Haystack.hello~&1~>.today[2]: 3
toTypescript: Haystack["hello"]["today"][2]

toString:     Haystack.hello~&1~>.today: [1, 2, 3]
toTypescript: Haystack["hello"]["today"]

toString:     Haystack.hello~&1~>: { today: [1, 2, 3]; }
toTypescript: Haystack["hello"]

toString:     Haystack.hello: { world: string; } & { today: [1, 2, 3]; }
toTypescript: Haystack["hello"]

toString:     Haystack.frog: GenericInterface<{ table: "block"; }, { value: string; }>
toTypescript: Haystack["frog"]

toString:     Haystack.frog: GenericInterface
toTypescript: Haystack["frog"]

toString:     Haystack.frog.constrained: { table: "block"; }
toTypescript: Haystack["frog"]["constrained"]

toString:     Haystack.frog.constrained.table: "block"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog.constrained.table: "block"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog.constrained: { table: "block"; }
toTypescript: Haystack["frog"]["constrained"]

toString:     Haystack.frog.defaulted: { value: string; }
toTypescript: Haystack["frog"]["defaulted"]

toString:     Haystack.frog.defaulted.value: string
toTypescript: Haystack["frog"]["defaulted"]["value"]

toString:     Haystack.frog.defaulted.value: string
toTypescript: Haystack["frog"]["defaulted"]["value"]

toString:     Haystack.frog.defaulted: { value: string; }
toTypescript: Haystack["frog"]["defaulted"]

toString:     Haystack.frog: GenericInterface
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~>: GenericInterface
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>: ParamWithConstraint
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>: Obj1
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table: "block" | "collection" | "activity"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table~|0~>: "block"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table~|0~>: "block"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table~|1~>: "collection"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table~|1~>: "collection"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table~|2~>: "activity"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table~|2~>: "activity"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.table: "block" | "collection" | "activity"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["method"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["method"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>~(param)~>: Obj2
toTypescript: Parameters<Haystack["frog"]["method"]>[0]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>~(param)~>.object2: true
toTypescript: Parameters<Haystack["frog"]["method"]>[0]["object2"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>~(param)~>.object2: true
toTypescript: Parameters<Haystack["frog"]["method"]>[0]["object2"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>~(param)~>: Obj2
toTypescript: Parameters<Haystack["frog"]["method"]>[0]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>.(): Obj2
toTypescript: ReturnType<Haystack["frog"]["method"]>

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>.().object2: true
toTypescript: ReturnType<Haystack["frog"]["method"]>["object2"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>.().object2: true
toTypescript: ReturnType<Haystack["frog"]["method"]>["object2"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>.(): Obj2
toTypescript: ReturnType<Haystack["frog"]["method"]>

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method~asFunction~>: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["method"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>.method: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["method"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>~constraintType~>: Obj1
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithConstraint>: ParamWithConstraint
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithDefault>: ParamWithDefault
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithDefault>~defaultType~>: Obj2
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithDefault>~defaultType~>.object2: true
toTypescript: Haystack["frog"]["object2"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithDefault>~defaultType~>.object2: true
toTypescript: Haystack["frog"]["object2"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithDefault>~defaultType~>: Obj2
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~><ParamWithDefault>: ParamWithDefault
toTypescript: Haystack["frog"]

toString:     Haystack.frog~instantiatedFrom~>.constrained: ParamWithConstraint
toTypescript: Haystack["frog"]["constrained"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>: Obj1
toTypescript: Haystack["frog"]["constrained"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table: "block" | "collection" | "activity"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table~|0~>: "block"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table~|0~>: "block"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table~|1~>: "collection"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table~|1~>: "collection"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table~|2~>: "activity"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table~|2~>: "activity"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.table: "block" | "collection" | "activity"
toTypescript: Haystack["frog"]["constrained"]["table"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["constrained"]["method"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["constrained"]["method"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>~(param)~>: Obj2
toTypescript: Parameters<Haystack["frog"]["constrained"]["method"]>[0]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>~(param)~>.object2: true
toTypescript: Parameters<Haystack["frog"]["constrained"]["method"]>[0]["object2"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>~(param)~>.object2: true
toTypescript: Parameters<Haystack["frog"]["constrained"]["method"]>[0]["object2"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>~(param)~>: Obj2
toTypescript: Parameters<Haystack["frog"]["constrained"]["method"]>[0]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>.(): Obj2
toTypescript: ReturnType<Haystack["frog"]["constrained"]["method"]>

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>.().object2: true
toTypescript: ReturnType<Haystack["frog"]["constrained"]["method"]>["object2"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>.().object2: true
toTypescript: ReturnType<Haystack["frog"]["constrained"]["method"]>["object2"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>.(): Obj2
toTypescript: ReturnType<Haystack["frog"]["constrained"]["method"]>

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method~asFunction~>: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["constrained"]["method"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>.method: (param: { object2: true; }) => { object2: true; }
toTypescript: Haystack["frog"]["constrained"]["method"]

toString:     Haystack.frog~instantiatedFrom~>.constrained~constraintType~>: Obj1
toTypescript: Haystack["frog"]["constrained"]

toString:     Haystack.frog~instantiatedFrom~>.constrained: ParamWithConstraint
toTypescript: Haystack["frog"]["constrained"]

toString:     Haystack.frog~instantiatedFrom~>.defaulted: ParamWithDefault
toTypescript: Haystack["frog"]["defaulted"]

toString:     Haystack.frog~instantiatedFrom~>.defaulted~defaultType~>: Obj2
toTypescript: Haystack["frog"]["defaulted"]

toString:     Haystack.frog~instantiatedFrom~>.defaulted~defaultType~>.object2: true
toTypescript: Haystack["frog"]["defaulted"]["object2"]

toString:     Haystack.frog~instantiatedFrom~>.defaulted~defaultType~>.object2: true
toTypescript: Haystack["frog"]["defaulted"]["object2"]

toString:     Haystack.frog~instantiatedFrom~>.defaulted~defaultType~>: Obj2
toTypescript: Haystack["frog"]["defaulted"]

toString:     Haystack.frog~instantiatedFrom~>.defaulted: ParamWithDefault
toTypescript: Haystack["frog"]["defaulted"]

toString:     Haystack.frog~instantiatedFrom~>: GenericInterface
toTypescript: Haystack["frog"]

toString:     Haystack.frog<<0>>: { table: "block"; }
toTypescript: Haystack["frog"]

toString:     Haystack.frog<<0>>.table: "block"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog<<0>>.table: "block"
toTypescript: Haystack["frog"]["table"]

toString:     Haystack.frog<<0>>: { table: "block"; }
toTypescript: Haystack["frog"]

toString:     Haystack.frog<<1>>: { value: string; }
toTypescript: Haystack["frog"]

toString:     Haystack.frog<<1>>.value: string
toTypescript: Haystack["frog"]["value"]

toString:     Haystack.frog<<1>>.value: string
toTypescript: Haystack["frog"]["value"]

toString:     Haystack.frog<<1>>: { value: string; }
toTypescript: Haystack["frog"]

toString:     Haystack.frog: GenericInterface<{ table: "block"; }, { value: string; }>
toTypescript: Haystack["frog"]

toString:     T: Haystack
toTypescript: T"
`;

exports[`visitDepthFirst 2`] = `
"
"
`;
